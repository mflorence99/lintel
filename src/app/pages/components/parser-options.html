<div [formGroup]="parserOptionsForm" class="parserOptions">
  <div class="option">
    <div class="tag">.ecmaVersion</div>

    <lintel-singleselector
      [enabled]="lintel.isEnabled"
      [options]="makeOptionsForSingleselector('parserOptions.properties.ecmaVersion')"
      [placeholder]="configs.extension.parserOptions?.ecmaVersion || ' '"
      formControlName="ecmaVersion"
    >
    </lintel-singleselector>

    <div class="desc">
      {{ schema.properties.parserOptions.properties.ecmaVersion.description }}
    </div>
  </div>

  <div class="option">
    <div class="tag">.sourceType</div>

    <lintel-singleselector
      [enabled]="lintel.isEnabled"
      [options]="makeOptionsForSingleselector('parserOptions.properties.sourceType')"
      [placeholder]="configs.extension.parserOptions?.sourceType || ' '"
      formControlName="sourceType"
    >
    </lintel-singleselector>

    <div class="desc">
      {{ schema.properties.parserOptions.properties.sourceType.description }}
    </div>
  </div>
</div>
