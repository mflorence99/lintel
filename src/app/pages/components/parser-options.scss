:host {
  display: block;
}

.parserOptions {
  display: grid;
  grid-column-gap: 0.5rem;
  grid-row-gap: 1rem;
  grid-template-columns: repeat(auto-fit, 20.5rem);

  .option {
    align-content: start;
    display: grid;
    grid-gap: 0.25rem;
    grid-template-columns: 1fr;

    .desc {
      color: var(--vscode-descriptionForeground);
      line-height: 125%;
    }

    .tag {
      color: var(--vscode-textPreformat-foreground);
      font-family: monospace;
    }
  }
}
