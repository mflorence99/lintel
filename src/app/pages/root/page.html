<mat-tree [dataSource]="dataSource" [treeControl]="treeControl">

  <mat-tree-node *matTreeNodeDef="let node" matTreeNodePadding>

    <span (click)="select(node.fileName)" class="leaf" [ngClass]="{selected: node.fileName === selection.fileName}">
      {{node.fileName}}
    </span>

  </mat-tree-node>

  <mat-tree-node *matTreeNodeDef="let node; when: hasChild" matTreeNodePadding>

    <mat-icon class="mat-icon-rtl-mirror" matTreeNodeToggle>
      {{treeControl.isExpanded(node) ? 'expand_more' : 'chevron_right'}}
    </mat-icon>

    <span (click)="select(node.fileName)" [ngClass]="{selected: node.fileName === selection.fileName}">
      {{node.fileName}}
    </span>

  </mat-tree-node>

</mat-tree>
