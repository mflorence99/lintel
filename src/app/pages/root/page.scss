:host {
  height: 100%;
}

.mat-tree-node {
  cursor: pointer;
}

.example-tree-invisible {
  display: none;
}

.example-tree ul,
.example-tree li {
  list-style-type: none;
  margin-bottom: 0;
  margin-top: 0;
}
