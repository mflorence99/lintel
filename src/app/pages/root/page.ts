import { ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { ConfigsState } from '../../state/configs';
import { ElementRef } from '@angular/core';
import { FilesState } from '../../state/files';
import { Params } from '../../services/params';
import { RulesState } from '../../state/rules';
import { SchemaState } from '../../state/schema';
import { SelectionState } from '../../state/selection';

/**
 * Lintel Root
 */

@Component({
  changeDetection: ChangeDetectionStrategy.Default,
  selector: 'lintel-root',
  templateUrl: 'page.html',
  styleUrls: ['page.scss']
})

export class RootPageComponent {

  /** ctor */
  constructor(public configs: ConfigsState,
              public files: FilesState,
              private host: ElementRef,
              public params: Params,
              public rules: RulesState,
              public schema: SchemaState,
              public selection: SelectionState) {
    // TODO: must do files first
    this.files.initialize();
    this.configs.initialize();
    this.rules.initialize();
    this.schema.initialize();
  }

  /** Scroll to top */
  scrollToTop(): void {
    const scroller = this.host.nativeElement.querySelector('#scroller');
    if (scroller)
      scroller.scrollTo({ top: 0, left: 0, behavior: 'smooth' });
  }

}
