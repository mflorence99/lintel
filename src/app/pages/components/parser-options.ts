import { ConfigsState } from '../../state/configs';
import { DestroyService } from '../../services/destroy';
import { ExtensionsState } from '../../state/extensions';
import { LintelState } from '../../state/lintel';
import { SchemaState } from '../../state/schema';
import { SelectionState } from '../../state/selection';
import { SingleselectorOptions } from '../../components/singleselector';
import { Utils } from '../../services/utils';

import { Actions } from '@ngxs/store';
import { ChangeDetectionStrategy } from '@angular/core';
import { ChangeDetectorRef } from '@angular/core';
import { Component } from '@angular/core';
import { FormBuilder } from '@angular/forms';
import { FormGroup } from '@angular/forms';
import { OnInit } from '@angular/core';

import { filter } from 'rxjs/operators';
import { takeUntil } from 'rxjs/operators';

/**
 * Parser options is complicated, so separate it from general.ts
 */

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [DestroyService],
  selector: 'lintel-parser-options',
  templateUrl: 'parser-options.html',
  styleUrls: ['parser-options.scss']
})
export class ParserOptionsComponent implements OnInit {
  parserOptionsForm: FormGroup;

  private underConstruction = true;

  /** ctor */
  constructor(
    private actions$: Actions,
    private cdf: ChangeDetectorRef,
    public configs: ConfigsState,
    private destroy$: DestroyService,
    public extensions: ExtensionsState,
    private formBuilder: FormBuilder,
    public lintel: LintelState,
    public schema: SchemaState,
    public selection: SelectionState,
    private utils: Utils
  ) {
    this.parserOptionsForm = this.formBuilder.group({
      ecmaVersion: null,
      sourceType: null
    });
  }

  /** Options from enum */
  makeOptionsForSingleselector(nm: string): SingleselectorOptions {
    const options = eval(`this.schema.properties.${nm}.enum`);
    return options.map((option) => [option, option]);
  }

  /** When we're ready */
  ngOnInit(): void {
    this.handleActions$();
    this.handleValueChanges$();
    this.rebuildControls();
  }

  // private methods

  private handleActions$(): void {
    this.actions$
      .pipe(
        filter(({ action, status }) => {
          return (
            this.utils.hasProperty(action, /^SelectionState\./) &&
            status === 'SUCCESSFUL'
          );
        }),
        takeUntil(this.destroy$)
      )
      .subscribe(() => {
        this.rebuildControls();
        this.cdf.markForCheck();
      });
  }

  private handleValueChanges$(): void {
    this.parserOptionsForm.valueChanges
      .pipe(
        filter(() => this.underConstruction),
        takeUntil(this.destroy$)
      )
      .subscribe((changes) => {
        this.configs.changeConfiguration({ parserOptions: changes });
      });
  }

  private rebuildControls(): void {
    this.underConstruction = false;
    this.parserOptionsForm.patchValue(
      {
        ecmaVersion:
          this.configs.configuration.parserOptions?.ecmaVersion ?? null,
        sourceType: this.configs.configuration.parserOptions?.sourceType ?? null
        // TODO: we don't know why { emitEvent: false } doesn't work
      },
      { emitEvent: false }
    );
    this.underConstruction = true;
  }
}
