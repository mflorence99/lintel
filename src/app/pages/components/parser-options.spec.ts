import { ParserOptionsComponent } from './parser-options';

import { prepare } from '../page.spec';

import 'jest-extended';

import { ComponentFixture } from '@angular/core/testing';
import { TestBed } from '@angular/core/testing';

describe('ParserOptionsComponent', () => {
  let component: ParserOptionsComponent;
  let fixture: ComponentFixture<ParserOptionsComponent>;

  beforeEach(() => {
    prepare();
    fixture = TestBed.createComponent(ParserOptionsComponent);
    component = fixture.componentInstance;
    component.ngOnInit();
  });

  test('makeOptionsForSingleselector', () => {
    component.selection.select({
      fileName: '/home/mflorence99/lintel/package.json'
    });
    const options = component.makeOptionsForSingleselector(
      'parserOptions.properties.ecmaVersion'
    );
    expect(options.find((option) => option[0] === 2020)).toBeTruthy();
  });

  test('ngOnInit', () => {
    component.selection.select({
      fileName: '/home/mflorence99/lintel/package.json'
    });
    component.ngOnInit();
    const changes = { ecmaVersion: 6 };
    component.parserOptionsForm.patchValue(changes);
    expect(component.configs.configuration.parserOptions).toEqual(
      expect.objectContaining(changes)
    );
  });
});
