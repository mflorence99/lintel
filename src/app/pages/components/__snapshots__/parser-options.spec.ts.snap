// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParserOptionsComponent snapshot 1`] = `
<lintel-parser-options
  __ngContext__={[Function LRootView]}
  actions$={[Function Actions]}
  cdf={[Function ViewRef]}
  configs={[Function ConfigsState]}
  destroy$={[Function DestroyService]}
  extensions={[Function ExtensionsState]}
  formBuilder={[Function FormBuilder]}
  lintel={[Function LintelState]}
  parserOptionsForm={[Function FormGroup]}
  schema={[Function SchemaState]}
  selection={[Function SelectionState]}
  underConstruction={[Function Boolean]}
  utils={[Function Utils]}
>
  <article
    class="parserOptions ng-untouched ng-pristine ng-valid"
  >
    <div
      class="option"
    >
      <p
        class="tag"
      >
        .ecmaVersion
      </p>
      <lintel-singleselector
        class="ng-untouched ng-pristine ng-valid"
        formcontrolname="ecmaVersion"
      >
        <select
          autocomplete="off"
          tabindex="0"
        >
          <option
            value="null"
          >
            2020
          </option>
          
          <option
            value="3"
          >
             3 
          </option>
          <option
            value="5"
          >
             5 
          </option>
          <option
            value="6"
          >
             6 
          </option>
          <option
            value="2015"
          >
             2015 
          </option>
          <option
            value="7"
          >
             7 
          </option>
          <option
            value="2016"
          >
             2016 
          </option>
          <option
            value="8"
          >
             8 
          </option>
          <option
            value="2017"
          >
             2017 
          </option>
          <option
            value="9"
          >
             9 
          </option>
          <option
            selected="true"
            value="2018"
          >
             2018 
          </option>
          <option
            value="10"
          >
             10 
          </option>
          <option
            value="2019"
          >
             2019 
          </option>
          <option
            value="11"
          >
             11 
          </option>
          <option
            value="2020"
          >
             2020 
          </option>
          
        </select>
      </lintel-singleselector>
      <p
        class="desc"
      >
         Set to 3, 5 (default), 6, 7, 8, 9, 10 or 11 to specify the version of ECMAScript syntax you want to use. You can also set to 2015 (same as 6), 2016 (same as 7), 2017 (same as 8), 2018 (same as 9), 2019 (same as 10) or 2020 (same as 11) to use the year-based naming. 
      </p>
    </div>
    <div
      class="option"
    >
      <p
        class="tag"
      >
        .sourceType
      </p>
      <lintel-singleselector
        class="ng-untouched ng-pristine ng-valid"
        formcontrolname="sourceType"
      >
        <select
          autocomplete="off"
          tabindex="0"
        >
          <option
            value="null"
          >
            module
          </option>
          
          <option
            value="script"
          >
             script 
          </option>
          <option
            selected="true"
            value="module"
          >
             module 
          </option>
          
        </select>
      </lintel-singleselector>
      <p
        class="desc"
      >
         Set to "script" (default) or "module" if your code is in ECMAScript modules. 
      </p>
    </div>
  </article>
</lintel-parser-options>
`;
